src/lazy/text/parse_result.rs

```diff
@@ -0,0 +1,274 @@
+    // The input that being parsed when the error arose
+    input: TextBufferView<'data>,
+    // A human-friendly name for what the parser was working on when the error occurred
+    label: Option<Cow<'static, str>>,
+    // The nature of the error--what went wrong?
+    description: Option<Cow<'static, str>>,
+    // A backtrace of errors that occurred leading to this one.
+    // XXX: This is the most expensive part of error handling and is likely not very useful.
+    //      Consider removing it if it doesn't carry its weight.
+    backtrace: Vec<InvalidInputError<'data>>,
+    // The nom ErrorKind, which indicates which nom-provided parser encountered the error we're
+    // bubbling up.
+    nom_error_kind: Option<ErrorKind>,
+}
+
+impl<'data> InvalidInputError<'data> {
+    /// Constructs a new `IonParseError` from the provided `input` text.
+    pub(crate) fn new(input: TextBufferView<'data>) -> Self {
+        InvalidInputError {
+            input,
+            label: None,
+            description: None,
+            nom_error_kind: None,
+            backtrace: Vec::new(),
+        }
+    }
+
+    /// Constructs a new `IonParseError` from the provided `input` text and `description`.
+    pub(crate) fn with_label<D: Into<Cow<'static, str>>>(mut self, label: D) -> Self {
+        self.label = Some(label.into());
+        self
+    }
+
+    /// Constructs a new `IonParseError` from the provided `input` text and `description`.
+    pub(crate) fn with_description<D: Into<Cow<'static, str>>>(mut self, description: D) -> Self {
+        self.description = Some(description.into());
+        self
+    }
+
+    /// Constructs a new `IonParseError` from the provided `input` text and `description`.
+    pub(crate) fn with_nom_error_kind(mut self, nom_error_kind: ErrorKind) -> Self {
+        self.nom_error_kind = Some(nom_error_kind);
+        self
+    }
+
+    pub(crate) fn append_error(&mut self, error: InvalidInputError<'data>) {
+        self.backtrace.push(error)
+    }
+
+    /// Returns a reference to the `description` text, if any.
+    pub fn description(&self) -> Option<&str> {
+        self.description.as_deref()
+    }
+
+    pub fn label(&self) -> Option<&str> {
+        self.label.as_deref()
+    }
+
+    // TODO: Decide how to expose 'input'.
+}
+
+// impl<'data> From<InvalidInputError<'data>> for IonError {
+//     fn from(value: InvalidInputError) -> Self {
+//         dbg!(&value.backtrace);
+//         let mut message = String::from(value.description().unwrap_or("invalid text Ion syntax"));
+//         if let Some(label) = value.label {
+//             message.push_str(" while ");
+//             message.push_str(label.as_ref());
+//         }
+//         let position = Position::with_offset(value.input.offset()).with_length(value.input.len());
+//         let decoding_error = DecodingError::new(message).with_position(position);
+//         IonError::Decoding(decoding_error)
+//     }
+// }
```

Comment:
🗺️ I'll remove this obsolete code after the PR is reviewed. Pushing code as I write the PR tour comments causes some number of the comments to be discarded 😐.